delivery/jwt.py

```diff
@@ -0,0 +1,162 @@
+import enum
+import functools
+
+import Crypto.PublicKey.RSA
+import Crypto.Util.number
+import dacite
+import jwt
+
+import model.delivery
+
+JWT_KEY = 'bearer_token'
+
+
+class KeyType(enum.StrEnum):
+    RSA = 'RSA'
+    OCTET_SEQUENCE = 'oct' # used to represent symmetric keys
+
+
+class Use(enum.StrEnum):
+    SIGNATURE = 'sig'
+    ENCRYPTION = 'enc'
+
+
+class Algorithm(enum.StrEnum):
+    RS256 = 'RS256'
+    HS256 = 'HS256'
+
+
+@dataclasses.dataclass(frozen=True)
+class JSONWebKey:
+    kty: KeyType
+    use: Use
+    alg: Algorithm
+    kid: str
+
+    @property
+    def key(self) -> bytes:
+        # has to be defined by derived classes
+        raise NotImplementedError
+
+    @staticmethod
+    def from_dict(data: dict) -> 'JSONWebKey':
+        algorithm = Algorithm(data.get('alg'))
+
+        class_by_algorithm = {
+            Algorithm.RS256: RSAPublicKey,
+            Algorithm.HS256: SymmetricKey,
+        }
+
+        return dacite.from_dict(
+            data_class=class_by_algorithm.get(algorithm),
+            data=data,
+            config=dacite.Config(
+                cast=[enum.Enum],
+            ),
+        )
+
+    @staticmethod
+    def from_signing_cfg(signing_cfg: model.delivery.SigningCfg) -> 'JSONWebKey':
+        algorithm = Algorithm(signing_cfg.algorithm().upper())
+        use = Use.SIGNATURE
+        kid = signing_cfg.id()
+
+        if algorithm == Algorithm.RS256:
+            public_key = Crypto.PublicKey.RSA.import_key(signing_cfg.public_key())
+
+            return RSAPublicKey(
+                use=use,
+                kid=kid,
+                n=encodeBase64urlUInt(public_key.n),
+                e=encodeBase64urlUInt(public_key.e),
+            )
+        elif algorithm == Algorithm.HS256:
+            return SymmetricKey(
+                use=use,
+                kid=kid,
+                k=encodeBase64url(signing_cfg.secret().encode('utf-8')),
+            )
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class RSAPublicKey(JSONWebKey):
+    n: str # modulus (Base64urlUInt encoded)
+    e: str # exponent (Base64urlUInt encoded)
+    kty: KeyType = KeyType.RSA
+    alg: Algorithm = Algorithm.RS256
+
+    @property
+    def key(self) -> bytes:
+        return Crypto.PublicKey.RSA.construct(
+            rsa_components=(
+                decodeBase64urlUInt(self.n),
+                decodeBase64urlUInt(self.e),
+            ),
+        ).export_key(format='PEM')
+
+
+@dataclasses.dataclass(frozen=True, kw_only=True)
+class SymmetricKey(JSONWebKey):
+    k: str # key value (base64url encoded)
+    kty: KeyType = KeyType.OCTET_SEQUENCE
+    alg: Algorithm = Algorithm.HS256
+
+    @property
+    def key(self) -> bytes:
+        return decodeBase64url(self.k)
+
+
+def encodeBase64url(b: bytes) -> str:
+    url_encoding = base64.urlsafe_b64encode(b)
+
+    return url_encoding.rstrip(b'=').decode('utf-8')
+
+
+def decodeBase64url(s: str) -> bytes:
+    return base64.urlsafe_b64decode(s + '==')
```

Comment:
why adding padding chars without checking str-length?
IIRC. those are used to ensure the str-length is a multiple of "4".

Reply:
According to the [documentation](https://docs.python.org/3/library/base64.html#base64.b64decode), an extra padding is being truncated prior to padding check:
> If validate is False (the default), characters that are neither in the normal base-64 alphabet nor the alternative alphabet are discarded prior to the padding check.

Hence, adding the maximum possible padding and letting `b64decode` check on the correct padding will just work fine.

Reply:
I only I had known that earlier some years back [here](https://github.com/gardener/cc-utils/blob/master/oci/client.py#L34-L45) 🙈